File: ios/Classes/Tasks/OfflineMap/OfflineMapTaskController.swift
//
// Created by Valentin Grigorean on 22.06.2022.
//

import Foundation
import ArcGIS

class OfflineMapTaskController {
    private let channel: FlutterMethodChannel
    private let messenger: FlutterBinaryMessenger
    private var offlineMapTasks: [Int: AGSOfflineMapTask] = [:]
    private var offlineMapCredentials: [Int: AGSCredential] = [:]
    private var jobs: [Int: GenerateOfflineMapJobController] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        channel = FlutterMethodChannel(name: "plugins.flutter.io/arcgis_channel/offline_map_task", binaryMessenger: messenger)
        channel.setMethodCallHandler(handle)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "createOfflineMapTask":
            createOfflineMapTask(data: call.arguments as! [String: Any])
            result(nil)
            break
        case "destroyOfflineMapTask":
            offlineMapTasks.removeValue(forKey: call.arguments as! Int)
            offlineMapCredentials.removeValue(forKey: call.arguments as! Int)
            result(nil)
            break
        case "defaultGenerateOfflineMapParameters":
            defaultGenerateOfflineMapParameters(data: call.arguments as! [String: Any], result: result)
            break
        case "destroyGenerateOfflineMapJob":
            jobs.removeValue(forKey: call.arguments as! Int)
            result(nil)
            break
        case "generateOfflineMap":
            generateOfflineMap(data: call.arguments as! [String: Any], result: result)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func generateOfflineMap(data: [String: Any], result: @escaping FlutterResult) {
        let id = data["id"] as! Int
        let jobId = data["jobId"] as! Int
        let parameters = AGSGenerateOfflineMapParameters(data: data["parameters"] as! [String: Any])
        let downloadDirectory = data["downloadDirectory"] as! String

        let offlineMapTask = offlineMapTasks[id]!
        let offlineMapJob = offlineMapTask.generateOfflineMapJob(with: parameters, downloadDirectory: URL(string: downloadDirectory)!)
        let jobController = GenerateOfflineMapJobController(messenger: messenger, id: jobId, offlineMapJob: offlineMapJob)
        jobs[jobId] = jobController
        result(nil)
    }

    private func defaultGenerateOfflineMapParameters(data: [String: Any], result: @escaping FlutterResult) {
        let id = data["id"] as! Int
        let offlineMapTask = offlineMapTasks[id]!
        let areaOfInterest = AGSGeometry.fromFlutter(data: data["areaOfInterest"] as! [String: Any])!
        let minScale = data["minScale"] as? Double
        let maxScale = data["maxScale"] as? Double

        if minScale == nil {
            offlineMapTask.defaultGenerateOfflineMapParameters(withAreaOfInterest: areaOfInterest, completion: { (parameters, error) in
                if let error = error {
                    result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(parameters?.toJSONFlutter())
                }
            })
        } else {
            offlineMapTask.defaultGenerateOfflineMapParameters(withAreaOfInterest: areaOfInterest, minScale: minScale!, maxScale: maxScale!, completion: { (parameters, error) in
                if let error = error {
                    result(FlutterError(code: "ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(parameters?.toJSONFlutter())
                }
            })
        }
    }

    private func createOfflineMapTask(data: [String: Any]) {
        var offlineMapTask: AGSOfflineMapTask
        if let map = data["map"] as? Dictionary<String, Any> {
            offlineMapTask = AGSOfflineMapTask(onlineMap: AGSMap(data: map))
        } else if let portalItem = data["portalItem"] as? Dictionary<String, Any> {
            offlineMapTask = AGSOfflineMapTask(portalItem: AGSPortalItem(data: portalItem))
        } else {
            fatalError("Invalid offline map task")
        }

        let id = data["id"] as! Int
        offlineMapTasks[id] = offlineMapTask
        if let credentials = data["credentials"] as? Dictionary<String, Any> {
            offlineMapCredentials[id] = AGSCredential(data: credentials)
        }
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/tasks/offlinemap/OfflineMapTaskController.java
package com.valentingrigorean.arcgis_maps_flutter.tasks.offlinemap;

import androidx.annotation.NonNull;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.geometry.Geometry;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.MobileMapPackage;
import com.esri.arcgisruntime.portal.PortalItem;
import com.esri.arcgisruntime.tasks.offlinemap.GenerateOfflineMapJob;
import com.esri.arcgisruntime.tasks.offlinemap.GenerateOfflineMapParameters;
import com.esri.arcgisruntime.tasks.offlinemap.GenerateOfflineMapResult;
import com.esri.arcgisruntime.tasks.offlinemap.OfflineMapTask;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class OfflineMapTaskController implements MethodChannel.MethodCallHandler {
    private final MethodChannel channel;
    private final BinaryMessenger messenger;
    private final Map<Integer, OfflineMapTask> offlineMapTasks = new HashMap<>();
    private final Map<Integer, ArrayList<GenerateOfflineMapJobController>> generateOfflineMapJobControllers = new HashMap<>();
    private final Map<Integer, GenerateOfflineMapJobController> generateOfflineMapJobControllersById = new HashMap<>();

    public OfflineMapTaskController(BinaryMessenger messenger) {
        this.messenger = messenger;
        channel = new MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/offline_map_task");
        channel.setMethodCallHandler(this);
    }

    public void dispose() {
        channel.setMethodCallHandler(null);
    }

    @Override
    protected void finalize() throws Throwable {
        dispose();
        super.finalize();
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "createOfflineMapTask":
                createOfflineMapTask(call.arguments());
                result.success(null);
                break;
            case "destroyOfflineMapTask":
                offlineMapTasks.remove(call.arguments());
                final ArrayList<GenerateOfflineMapJobController> controllers = generateOfflineMapJobControllers.remove(call.arguments());
                for (GenerateOfflineMapJobController controller : controllers) {
                    controller.dispose();
                }
                result.success(null);
                break;
            case "defaultGenerateOfflineMapParameters":
                defaultGenerateOfflineMapParameters(call.arguments(), result);
                break;
            case "destroyGenerateOfflineMapJob":
                final int id = call.arguments();
                final GenerateOfflineMapJobController controller = generateOfflineMapJobControllersById.remove(id);
                if (controller != null) {
                    controller.dispose();
                }
                generateOfflineMapJobControllers.forEach((key, value) -> value.remove(controller));
                result.success(null);
                break;
            case "generateOfflineMap":
                generateOfflineMap(call.arguments(), result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void generateOfflineMap(Map<?, ?> data, MethodChannel.Result result) {
        final int id = (int) data.get("id");
        final int jobId = (int) data.get("jobId");
        final GenerateOfflineMapParameters parameters = ConvertOfflineMap.toGenerateOfflineMapParameters(data.get("parameters"));
        final OfflineMapTask offlineMapTask = offlineMapTasks.get(id);
        final String downloadDirectory = (String) data.get("downloadDirectory");
        final GenerateOfflineMapJob offlineMapJob = offlineMapTask.generateOfflineMap(parameters, downloadDirectory);
        final GenerateOfflineMapJobController controller = new GenerateOfflineMapJobController(messenger, jobId, offlineMapJob);
        generateOfflineMapJobControllers.get(id).add(controller);
        generateOfflineMapJobControllersById.put(jobId, controller);
        result.success(null);
    }

    private void defaultGenerateOfflineMapParameters(Map<?, ?> data, MethodChannel.Result result) {
        final int id = (int) data.get("id");
        final OfflineMapTask offlineMapTask = offlineMapTasks.get(id);
        final Geometry areaOfInterest = Convert.toGeometry(data.get("areaOfInterest"));
        final Object minScale = data.get("minScale");
        final Object maxScale = data.get("maxScale");
        ListenableFuture<GenerateOfflineMapParameters> future;
        if (minScale == null) {
            future = offlineMapTask.createDefaultGenerateOfflineMapParametersAsync(areaOfInterest);
        } else {
            future = offlineMapTask.createDefaultGenerateOfflineMapParametersAsync(areaOfInterest, (double) minScale, (double) maxScale);
        }

        future.addDoneListener(() -> {
            try {
                final GenerateOfflineMapParameters parameters = future.get();
                result.success(ConvertOfflineMap.generateOfflineMapParametersToJson(parameters));
            } catch (Exception e) {
                result.error("defaultGenerateOfflineMapParameters", e.getMessage(), null);
            }
        });
    }

    private void createOfflineMapTask(Map<?, ?> data) {
        OfflineMapTask offlineMapTask;
        final Object map = data.get("map");
        final Object portalItem = data.get("portalItem");
        if (map != null) {
            ArcGISMap arcGISMap = Convert.toArcGISMap(map);
            offlineMapTask = new OfflineMapTask(arcGISMap);
        } else if (portalItem != null) {
            final PortalItem nativePortalItem = Convert.toPortalItem(portalItem);
            offlineMapTask = new OfflineMapTask(nativePortalItem);
        } else {
            throw new IllegalArgumentException("Map or PortalItem is required");
        }

        final int id = (int) data.get("id");
        offlineMapTasks.put(id, offlineMapTask);
        final Object credentials = data.get("credentials");
        if (credentials != null) {
            offlineMapTask.setCredential(Convert.toCredentials(credentials));
        }
        generateOfflineMapJobControllers.put(id, new ArrayList<>());
    }
}
